include/data_buffer.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+#include <string>
+#include <vector>
+
+namespace reinforcement_learning { namespace utility {
+
+/*
+     * Data buffer used for serialized messages.  Data buffer
+     * consists of 2 parts.  1) preamble 2) body
+     *
+     * body: can start somewhere in the middle and end before the
+     * buffer ends.
+     * preamble: is n bytes immediately before the body.
+     * preamble is usually 8 bytes and sometimes zero if there is no preamble
+     */
+class data_buffer {
+public:
+  using value_type = unsigned char;
+
+  explicit data_buffer(size_t body_size = 1024);
+
+  // Get a pointer to beginning of preamble
+  value_type *preamble_begin();
+
+  constexpr size_t preamble_size() const;
+
+  // Return pointer to beginning of buffer, starting at offset from the real beginning.
+  value_type *body_begin();
+
+  // Body size (does not include the preamble.)
+  size_t body_filled_size() const;
+
+  // How large can we make the Body? (does not include the preamble.)
+  size_t body_capacity() const;
+
+  // Offset for beginning of the body region from the beginning of the buffer
+  size_t get_body_beginoffset() const;
```

Comment:
it feels like there should be a _ between begin and offset

Reply:
I went back and forth on naming this funtion.  The competing elements are 1) body is an entity in itself. begin_offset is an idea pertaining to body.  vs 2) get_body_begin_offset just seems cleaner.  Ultimately the reason I wend with 1 is that this class is particularly confusing around the idea of body and preamble and it made more sense to make that less confusing than make this name seem cleaner.